docs/guides/hybrid-graphics.md

```diff
@@ -77,6 +79,65 @@ If you experience system freezes, then the laptop's fans becoming loud, before t
 
 `glxinfo | grep "OpenGL renderer"` should show an Intel GPU. Running programs with `DRI_PRIME=1` will make them render on your AMD GPU (some things do this automatically). You will get more battery time now as your AMD GPU can be turned off when not needed.
 
+### NixOS
+
+OBS! These instructions assumes a systemd bootloader. If using something else you will need to make some minor tweaks to assure the .efi file ends up in correct dir and gets set as default.
+
+1.  Create a new file in `/etc/nixos/` called `hybrid_graphics.nix`:
+
+    ```plain
+    { config, pkgs, ... }:
+    let
+    hybrid-graphics = pkgs.stdenv.mkDerivation rec {
+        name = "hybrid-graphics-1.0";
+
+        src = pkgs.fetchFromGitHub {
+        owner = "Redecorating";
+        repo = "apple_set_os-loader";
+        rev = "r33.9856dc4";
+        sha256 = "hvwqfoF989PfDRrwU0BMi69nFjPeOmSaD6vR6jIRK2Y=";
+        };
+
+        buildInputs = [ pkgs.gnu-efi ];
+
+        buildPhase = ''
+        substituteInPlace Makefile --replace "/usr" '$(GNU_EFI)'
+        export GNU_EFI=${pkgs.gnu-efi}
+        make
+        '';
+
+        installPhase = ''
+        install -D bootx64_silent.efi $out/bootx64.efi
+        '';
+    };
+    in
+    {
+    system.activationScripts.hybrid-graphics = {
+        text = ''
+        mkdir -p /boot/efi/EFI/hybrid_graphics
+        cp ${hybrid-graphics}/bootx64.efi /boot/efi/EFI/hybrid_graphics/bootx64.efi
+        cp /boot/efi/EFI/BOOT/BOOTX64.EFI /boot/efi/EFI/BOOT/bootx64_original.efi
+        '';
+    };
+
+    environment.etc."modprobe.d/apple-gmux.conf".text = ''
+    # Enable the iGPU by default if present
+    options apple-gmux force_igd=y
+    '';
+
+    environment.systemPackages = with pkgs; [ hybrid-graphics ];
+    }
+    ```
+
+2.  Add it to the imports in `/etc/nixos/configuration.nix`.
+
+3.  ```sh
+    sudo nixos-rebuild switch
+    ```
+4.  Use `efibootmgr` to create a new default boot entry point: `/boot/efi/EFI/hybrid_graphics/bootx64.efi` (see https://nixos.wiki/wiki/Bootloader for more details). It will then automatically load `/boot/efi/EFI/BOOT/bootx64_original.efi` which is a copy of the main loader (in this case systemd), so carefully check to make sure that file has been generated correctly. If not then adjust the relevant lines in `hybrid_graphics.nix`.
```

Comment:
Hmm... Not sure how you mean? When installing systemd then `/boot/efi/EFI/systemd/bootx64.efi` will be the primary entry right? And it makes `/boot/efi/EFI/BOOT/bootx64.efi` a copy of `/boot/efi/EFI/systemd/bootx64.efi` to use as fallback. 

So if we copy the hybrid gfx .efi to `/boot/efi/EFI/BOOT/bootx64.efi` we would overwrite that fallback, and we would still need to use efibootmgr to set `/boot/efi/EFI/BOOT/bootx64.efi` as new primary entry, or it would just boot the systemd one. So I thought it's more clear what's going on with a new dedicated hybrid gfx dir.

Maybe i misunderstood your comment?

Reply:
> Hmm... Not sure how you mean? When installing systemd then `/boot/efi/EFI/systemd/bootx64.efi` will be the primary entry right? And it makes `/boot/efi/EFI/BOOT/bootx64.efi` a copy of `/boot/efi/EFI/systemd/bootx64.efi` to use as fallback.

Yes, but apple firmware only uses the "fallback" location. The fallback location is really the default location.

> So if we copy the hybrid gfx .efi to `/boot/efi/EFI/BOOT/bootx64.efi` we would overwrite that fallback

Thats what we want to do, the "fallback" is really what is booted

> and we would still need to use efibootmgr to set `/boot/efi/EFI/BOOT/bootx64.efi` as new primary entry, or it would just boot the systemd one.

It is set by path, not binary content, so you would not need to update the entry.

> So I thought it's more clear what's going on with a new dedicated hybrid gfx dir.

apple_set_os-loader will never be loaded because its not on the default boot path. apple-set-os-loader needs to be at `EFI/BOOT/bootx64.efi`, the real bootloader needs to be at `EFI/BOOT/bootx64_original`, and the default entry does not need to be updated

Reply:
That's strange...On my computer it does boot just fine.. this is my current setup:

```
❯  efibootmgr 
BootCurrent: 0002
Timeout: 2 seconds
BootOrder: 0002,0001,0080
Boot0000* Windows Boot Manager	HD(1,GPT,b11e6afa-c9ca-457c-980b-7da5659607de,0x6,0x12c00)/File(\EFI\Microsoft\Boot\bootmgfw.efi)57494e444f5753000100000088000000780000004200430044004f0042004a004500430054003d007b00390064006500610038003600320063002d0035006300640064002d0034006500370030002d0061006300630031002d006600330032006200330034003400640034003700390035007d00000033000100000010000000040000007fff0400
Boot0001* Linux Boot Manager	HD(1,GPT,b11e6afa-c9ca-457c-980b-7da5659607de,0x30,0x96000)/File(\EFI\systemd\systemd-bootx64.efi)
Boot0002* Hybrid Graphics	HD(1,GPT,b11e6afa-c9ca-457c-980b-7da5659607de,0x6,0x12c00)/File(\EFI\hybrid_graphics\bootx64.efi)
Boot0080* 	PciRoot(0x0)/Pci(0x1b,0x0)/Pci(0x0,0x0)//HD(3,GPT,c0559b8e-11ea-7e41-bec0-9d72091747a5,0x5d32700,0x12c00)/File(\EFI\BOOT\BOOTX64.EFI)
Boot0081* Mac OS X	PciRoot(0x0)/Pci(0x1b,0x0)/Pci(0x0,0x0)//HD(3,GPT,de796cbd-1c32-42f4-99ef-768ebf98b860,0x5aa5000,0x254000)
Boot0082* 	PciRoot(0x0)/Pci(0x1b,0x0)/Pci(0x0,0x0)//HD(2,GPT,79297f36-ad21-49d8-b9e3-5278901ac838,0x12c06,0x5d1fa9b)/VenMedia(be74fcf7-0b7c-49f3-9147-01f4042e6842,0ee42abd8e12764482a16b0790947a29)/File(\ACEBDF43-C81C-49E1-9FDB-B877E0E5DAAD\System\Library\CoreServices\boot.efi)
```

Reply:
> That's strange...On my computer it does boot just fine.. this is my current setup:
> 
> ```
> ❯  efibootmgr 
> BootCurrent: 0002
> Timeout: 2 seconds
> BootOrder: 0002,0001,0080
> Boot0000* Windows Boot Manager	HD(1,GPT,b11e6afa-c9ca-457c-980b-7da5659607de,0x6,0x12c00)/File(\EFI\Microsoft\Boot\bootmgfw.efi)57494e444f5753000100000088000000780000004200430044004f0042004a004500430054003d007b00390064006500610038003600320063002d0035006300640064002d0034006500370030002d0061006300630031002d006600330032006200330034003400640034003700390035007d00000033000100000010000000040000007fff0400
> Boot0001* Linux Boot Manager	HD(1,GPT,b11e6afa-c9ca-457c-980b-7da5659607de,0x30,0x96000)/File(\EFI\systemd\systemd-bootx64.efi)
> Boot0002* Hybrid Graphics	HD(1,GPT,b11e6afa-c9ca-457c-980b-7da5659607de,0x6,0x12c00)/File(\EFI\hybrid_graphics\bootx64.efi)
> Boot0080* 	PciRoot(0x0)/Pci(0x1b,0x0)/Pci(0x0,0x0)//HD(3,GPT,c0559b8e-11ea-7e41-bec0-9d72091747a5,0x5d32700,0x12c00)/File(\EFI\BOOT\BOOTX64.EFI)
> Boot0081* Mac OS X	PciRoot(0x0)/Pci(0x1b,0x0)/Pci(0x0,0x0)//HD(3,GPT,de796cbd-1c32-42f4-99ef-768ebf98b860,0x5aa5000,0x254000)
> Boot0082* 	PciRoot(0x0)/Pci(0x1b,0x0)/Pci(0x0,0x0)//HD(2,GPT,79297f36-ad21-49d8-b9e3-5278901ac838,0x12c06,0x5d1fa9b)/VenMedia(be74fcf7-0b7c-49f3-9147-01f4042e6842,0ee42abd8e12764482a16b0790947a29)/File(\ACEBDF43-C81C-49E1-9FDB-B877E0E5DAAD\System\Library\CoreServices\boot.efi)
> ```

Are you booting it with GRUB/sd-boot or the apple bootloader. On my system the other EFI entries are ignored

Reply:
Aha.. now I get it! I tried rebooting while holding Option now to test and then it indeed goes by path, ignoring apple_set_os-loader. 

I rarely boot into Mac OS so I blessed the linux efi in order for it to auto boot. I only use Option-boot for getting into Mac OS on occation. So I never noticed that custom entries won't work when Option-booting into linux. Thanks for pointing that out.. I'll change it,

Reply:
@sharpenedblade I've updated the script and docs to make it use the standard path.

Reply:
It looks good, you can mark this as resolved